docs/dev/relay_add_pass.rst

```diff
@@ -0,0 +1,125 @@
+.. _relay-add-pass:
+
+Adding a Compiler Pass to Relay
+===============================
+
+Compiler passes can be used to collect data about Relay programs or transform
+them in various ways. The base class used to traverse programs is
+``ExprFunctor``. Depending on what you want to do, there are subclasses
+``ExprVisitor`` and ``ExprMutator`` that collect information or mutate the
+AST, respectively.
+
+At a high level, there are three key components to writing a pass:
+
+- Creating one or more C++ classes that traverse the program
+- Registering an API endpoint with the ``TVM_REGISTER_API`` macro that performs the pass
+- Wrapping the Python API hook in a neater interface
+
+In order to better understand the process of writing a pass, we will look at
+the constant folding pass (found in ``src/relay/pass/fold_constant.cc`` and
+in ``python/tvm/relay/ir_pass.py``) as a guide, because it is a relatively
+simple pass that incorporates both types of traversals.
+
+Background
+----------
+
+Constant folding involves evaluating expressions in the program that only
+involve constant values, then replacing those expressions with the result
+of evaluating them. The goal of this pass is to frontload all of the
+computations that we can. With that background, we now present an example of
+an ``ExprVisitor``, then an example of an ``ExprMutator``.
+
+Writing an Expression Visitor
+-----------------------------
+
```

Comment:
It might be good to mention that the recommended practice for writing Visitors is to have a separate public interface method (like `Check()` in the example) rather than using `VisitExpr()`.